Resolves gh-5: Implements a relay server for PhET-iO events via OSC

colinbdclark opened the pull request:
This pull request includes #2 and #4. This adds full support for converting PhET-iO events into OSC messages and sending them over an osc.js output to a remote server such as SuperCollider or Max.

colinbdclark committed:
Adds links to PhET.io, OSC, and osc.js.

colinbdclark committed:
Resolves gh-1: Adds testing infrastructure and basic failing test.

colinbdclark committed:
gh-3: Implements basic logic for converting primitive-valued event parameters.

colinbdclark committed:
gh-3: Adds support for converting object-valued event parameters.

colinbdclark committed:
gh-3: Adds support for composite events.

colinbdclark committed:
gh-3: Refactors conversion functions to avoid mutating shared data.

colinbdclark committed:
gh-3: Fixes linting errors.

colinbdclark committed:
gh-5: Implements basic structure for relay server. No tests yet.

colinbdclark committed:
Merge branch 'gh-3' into gh-5

* gh-3:
  gh-3: Fixes linting errors.

colinbdclark committed:
gh-5: Adds support for streaming all PhET events to OSC.

colinbdclark committed:
gh-5: Adds support for event filtering.

colinbdclark committed:
gh-5: Adds support for sending to SuperCollider.

colinbdclark committed:
gh-5: Fixes linting errors.

colinbdclark committed:
gh-5: Adds support for excluding parameters. New filtering tests.

colinbdclark committed:
gh-5: Fixes linting errors.

colinbdclark committed:
gh-5: Ensures UDPPort is not multiply bound. Adds build process.

colinbdclark committed:
gh-5: Adds support for address remapping.

colinbdclark committed:
gh-5: Fixes linting errors. Builds scripts on npm prepublish.

colinbdclark committed:
gh-5: Adds support for gpii-launcher. Improves startup logs.

colinbdclark committed:
gh-5: Adds documentation and support for global command installation.

colinbdclark committed:
gh-5: Fixes linting errors.

colinbdclark committed:
gh-5: Adds SuperCollider example and unhides the examples directory.

amb26 commented on README.md:
```diff
@@ -1,39 +1,124 @@
+
+Next, you'll need to instantiate the bridge component and bind it to the PhET sim's <code>simIFrameClient</code> object. This will cause all events emitted from the sim to be available across the bridge. Paste this code into the <code>onSimInitialized</code> event handler function that is included in your wrapper's call to <code>simIFrameClient.launchSim()</code>:
+
+    var oscBridge = phetosc.bridge();
```
A full snippet including the event handler might be helpful here

amb26 commented on README.md:
```diff
@@ -1,39 +1,124 @@
+The <code>phetosc.bridge</code> component provides a variety of configuration options, which are described in the sections below.
+
+### Filtering Events
+In many cases, only a subset of events are relevant for a sonification. The <code>phetioIDPatterns</code> option allows you to specify an array of regular expression patterns that will be matched against the <code>phetioID</code> of events. Only those events with a <code>phetioID</code> matching the specified list of regular expressions will be sent via OSC.
```
Comment could be clarified to explain that the patterns will be ORed together as per https://github.com/fluid-studios/phet-osc-bridge/pull/6/files#diff-8621dab8da352b7954438b801f34de0cR51 - that is, it is sufficient to match any one of the expressions

amb26 commented on src/js/phet-converter.js:
```diff
@@ -0,0 +1,181 @@
+};
+
+phetosc.converter.objectToMessages = function (addressPrefix, that, obj) {
+    if (!obj || Object.keys(obj) < 1) {
```
I imagine you meant to say here something like Object.keys(obj).length < 1, but a) checking for length === 0 seems a touch more idiomatic, and b) it's not clear what happens to objects which have more contents than this - it appears that the actual contents of "obj" are always ignored?

amb26 replied:
Sorry, ignore 2nd half of comment, I just scrolled down a little further on this tiny screen. Function body might be a touch clearer with 2nd half enclosed in "else"

amb26 commented on src/js/port.js:
```diff
@@ -0,0 +1,109 @@
+fluid.defaults("phetosc.port", {
+    gradeNames: "fluid.component",
+
+    portType: "fluid.mustBeOverridden",
```
Some docs on acceptable values would be helpful here

amb26 commented on tests/js/conversion-tester.js:
```diff
@@ -0,0 +1,103 @@
+});
+
+phetosc.conversionTester.loadFile = function (path, onLoaded, onError) {
+    var loader = typeof $ !== "undefined" ? phetosc.conversionTester.browserLoadFile :
```
Doesn't seem enormously safe since $ is often defined on the server (although clearly not by you, here) - how about fluid.contextAware.isBrowser()

amb26 commented on tests/js/phet-converter-tests.js:
```diff
@@ -0,0 +1,227 @@
+                    filter: {
+                        type: "phetosc.eventFilter",
+                        options: {
+                            phetioIDPatterns: "{withIDPatterns}.options.phetioIDPatterns"
```
Just an options distribution for this option required? I think the component skeleton is already present?

colinbdclark committed:
gh-5: Addresses code review feedback. Thanks, @amb26!

colinbdclark commented:
Thank you very much for the code review, @amb26. I have addressed all your review comments. Do you think we're ready to ship the first version of this now?

amb26 commented:
Useful utility - I think it looks good to :ship: !
